include/drivers/rtc.h

```diff
@@ -0,0 +1,262 @@
+#include <time.h>
+#include <device.h>
+// #include <dt-bindings/rtc/rtc.h>
+
+/**
+ * @brief RTC driver APIs
+ * @defgroup rtc_interface RTC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/** @brief Wake up timers enum. */
+enum rtc_wakeup_id {
+	RTC_WUT0,       /**< Wake Up timer 0. */
+	/* Add new item before RTC_WUT_NUM. */
+	RTC_WUT_NUM,    /**< Number of wut for the driver. */
+};
+
+/** @brief Alarms enum. */
+enum rtc_alarm_id {
+	RTC_ALARMA,     /**< Alarm A. */
+	RTC_ALARMB,     /**< Alarm B. */
+	/* Add new item before RTC_ALARM_NUM. */
+	RTC_ALARM_NUM,  /**< Number of alarms for the driver. */
+};
+
+/** @brief Specifies if the RTC Alarm is on date or Week day. */
+enum rtc_alarm_date_weekday {
+	RTC_ALARM_DATE_SEL,     /**< Alarm Date is selected, set parameter to a value in the 1-31 range. */
+	RTC_ALARM_WEEKDAY_SEL,  /**< Alarm Weekday is selected, set parameter to a value in the 0-6 range (days since Sunday). */
+};
+
+/** @brief Specifies the RTC Alarm Masks. */
+enum rtc_alarm_mask {
+	RTC_ALARM_MASK_NONE,            /**< No mask. */
+	RTC_ALARM_MASK_DATE_WEEKDAY,    /**< Mask on date or weekday. */
+	RTC_ALARM_MASK_HOURS,           /**< Mask on hours. */
+	RTC_ALARM_MASK_MIN,             /**< Mask on minutes. */
+	RTC_ALARM_MASK_SEC,             /**< Mask on seconds. */
+	RTC_ALARM_MASK_ALL,             /**< Mask on all. */
+};
+
+/** @brief Wake up timer structure. */
+struct rtc_wakeup {
+	uint32_t period;                                /**< Wake up timer period in milliseconds. */
+	void (*callback)(const struct device *dev);     /**< Callback function. Can be NULL. */
+};
+
+/** @brief Alarm structure.
+ * @param alarm_time Alarm time members
+ * @param alarm_date_weekday_sel Alarm selection: on Date or on WeekDay.
+ * 	If Alarm date is selected, use tm_mday [1 to 31] to set the alarm
+ *  If WeekDay date is selected, use tm_wday [0 to 6] to set the alarm
+ * @param callback	Callback function. Can be NULL.
+ */
+struct rtc_alarm {
+	struct tm alarm_time;           /**< Alarm time members. */
+	enum rtc_alarm_mask alarm_mask; /**< Alarm mask. */
+	/** Alarm selection: on Date or on WeekDay.
+	 * If Alarm date is selected, use tm_mday [1 to 31] to set the alarm
+	 * If WeekDay date is selected, use tm_wday [0 to 6] to set the alarm
+	 */
+	enum rtc_alarm_date_weekday alarm_date_weekday_sel;
+	void (*callback)(const struct device *dev);     /**< Callback function. Can be NULL. */
+};
+
+/**
+ * @brief RTC driver API
+ *
+ * This is the mandatory API any RTC driver needs to expose.
+ */
+__subsystem struct rtc_driver_api {
+	int (*set_time)(const struct device *dev, const struct tm date_time);
+	int (*get_time)(const struct device *dev, struct tm *date_time);
+	int (*set_wakeup_timer)(const struct device *dev, struct rtc_wakeup wut, enum rtc_wakeup_id wut_id);
+	int (*start_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
+	int (*stop_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
```

Comment:
Shouldn't wake up be separate API? Have external i2c/parallel/spi devices been considered?
You may have devices like https://datasheets.maximintegrated.com/en/ds/DS1371.pdf or https://ww1.microchip.com/downloads/en/DeviceDoc/MCP7940M-Low-Cost%20I2C-RTCC-with-SRAM-20002292C.pdf, connected via i2c, that may use the alarm API but does not have wake up functionality. On the other side it is possible, on some devices, to use watchdog as a wake up device, even though the device itself does not have RTC, only various timers.
Having "wake-up" functionality as mandatory means that a lot of i2c devices will just have -ENOTSUP added just because it is required.

Reply:
Some external RTC (like NXP PCF85063A) use this feature under the name "Countdown timer" or "programmable timer".
"wakeup timer" is the name used by ST.
I suggest keeping this feature in the RTC API because this timer is specific to the RTC and sometimes it'usefull for ultra low power periodic wake-up. The user is free to use them or not. In this way the API is generic.
Like for the other driver, if an RTC doesn't implement the wakeup timer fonctionnality, the driver can return -ENOTSUP with a log message:
		LOG_ERR("Wake up timer is not supported");
		return -ENOTSUP;

Reply:
>(...) In this way the API is generic. Like for the other driver, if an RTC doesn't implement the wakeup timer fonctionnality, the driver can return -ENOTSUP with a log message: LOG_ERR("Wake up timer is not supported"); return -ENOTSUP;

Problem with generic APIs that they bring a lot of stuff that you may not be using but will take up your flash anyway.
Separate wake-API would be more generic because it would allow to not only cover this specific RTC tied feature but the mentioned watchdog based or count down based, whatever it is called, when you are not interested in entire RTC thing.

Reply:
We can add a symbol in the Kconfig of the rtc

config RTC_EN_WUT
	bool "Enable the rtc wake up timer support"
	help
	  This option enables the rtc wake up timer API calls.
	  
And decide to enable or not the wake up timer API
__subsystem struct rtc_driver_api {
	int (*set_time)(const struct device *dev, const struct tm date_time);
	int (*get_time)(const struct device *dev, struct tm *date_time);
#ifdef CONFIG_RTC_EN_WUT
	int (*set_wakeup_timer)(const struct device *dev, struct rtc_wakeup wut, enum rtc_wakeup_id wut_id);
	int (*start_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
	int (*stop_wakeup_timer)(const struct device *dev, enum rtc_wakeup_id wut_id);
#endif
	int (*set_alarm)(const struct device *dev, struct rtc_alarm alarm, enum rtc_alarm_id alarm_id);
	int (*start_alarm)(const struct device *dev, enum rtc_alarm_id alarm_id);
	int (*stop_alarm)(const struct device *dev, enum rtc_alarm_id alarm_id);
};
idem for the wut data ...

In this way, an rtc can implement the wut or not without additional memory constraints ...

Reply:
I am not sure whether #ifdefing API interface is good idea, I know it is done at least with Flash API, but I remember that there have been some arguments against this approach.

Anyway, your proposal is actually reverse of I was thinking of.
I will describe thing I did some time ago to better let you understand what I have meant:
the device I  have in mind was controlling several other devices and conducting certain tests on them (that could take several minutes) at specified intervals and then would go to sleep; the device had been sleeping most of the time, when not doing its main job, to save power as it could be running on batteries (VRLA, AGM or GEL, on a power failure), and would be woken up by a watchdog at certain intervals (ranging from minutes to hours) depending on previous state. The second wake up channel, external, would be a communication device that would wake up the control device upon receiving request from base.
So both ways to get the device to life would be via wake up channel, one timed and one asynchronous interrupt, in both cases the device would check comm device for any messages to be processed.
As you can see such device would have to get entire API to just use the wake up portion of it.

Reply:
I agree with @de-nordic that this seems better as a separate API to implement in addition to the RTC API, rather than lumped in here with an #ifdef. It feels like it's polluting the core idea of what it means to be an RTC with a (common but specific) implementation type.

Reply:
Therefore, those 3 functions should be removed form rtc API